src/firewall/firewalld.rs

```diff
@@ -333,19 +335,29 @@ impl firewall::FirewallDriver for FirewallD {
             for port_tuple in old_port_forwarding_rules.iter() {
                 match port_tuple {
                     Value::Structure(s) => {
-                        let fields = s.clone().into_fields();
+                        let fields = s
+                            .try_clone()
+                            .map_err(|e| {
+                                NetavarkError::wrap(
+                                    "Invalid port forwarding rule encountered",
+                                    e.into(),
+                                )
+                            })?
+                            .into_fields();
                         if fields.len() != 4 {
                             return Err(NetavarkError::msg(
                                 "Port forwarding rule that was not a 4-tuple encountered",
                             ));
                         }
-                        let port_ip = match fields[3].clone() {
-                            Value::Str(s) => s.as_str().to_string(),
+                        let port_ip = match fields[3].try_clone() {
+                            Ok(Value::Str(s)) => s.as_str().to_string(),
                             _ => return Err(NetavarkError::msg("Port forwarding tuples must contain only strings, encountered a non-string object")),
                         };
                         debug!("IP string from firewalld is {}", port_ip);
                         if port_ip != ipv4 && port_ip != ipv6 {
-                            port_forwarding_rules.append(port_tuple.clone())?;
+                            port_forwarding_rules.append(port_tuple.try_clone().map_err(
+                                |e| NetavarkError::wrap("TODO: write a proper message", e.into()),
```

Comment:
@Luap99 @mheon please help decide how this should be written; I don't really have an idea.

Reply:
"Unable to clone port forwarding tuple" maybe?

Reply:
Sure, I'll wait for all checks to finish then push it.

Reply:
You can remove the clones above here actually

```diff
diff --git a/src/firewall/firewalld.rs b/src/firewall/firewalld.rs
index ab2dcaf..e6e4c2b 100644
--- a/src/firewall/firewalld.rs
+++ b/src/firewall/firewalld.rs
@@ -1,8 +1,8 @@
 use crate::error::{NetavarkError, NetavarkResult};
-use crate::firewall;
 use crate::network::internal_types;
 use crate::network::internal_types::{PortForwardConfig, TearDownNetwork, TeardownPortForward};
 use crate::network::types::PortMapping;
+use crate::{firewall, wrap};
 use core::convert::TryFrom;
 use log::{debug, info, warn};
 use std::collections::HashMap;
@@ -335,28 +335,21 @@ impl firewall::FirewallDriver for FirewallD {
             for port_tuple in old_port_forwarding_rules.iter() {
                 match port_tuple {
                     Value::Structure(s) => {
-                        let fields = s
-                            .try_clone()
-                            .map_err(|e| {
-                                NetavarkError::wrap(
-                                    "Invalid port forwarding rule encountered",
-                                    e.into(),
-                                )
-                            })?
-                            .into_fields();
+                        let fields = s.fields();
                         if fields.len() != 4 {
                             return Err(NetavarkError::msg(
                                 "Port forwarding rule that was not a 4-tuple encountered",
                             ));
                         }
-                        let port_ip = match fields[3].try_clone() {
-                            Ok(Value::Str(s)) => s.as_str().to_string(),
+                        let port_ip = match &fields[3] {
+                            Value::Str(s) => s.as_str().to_string(),
                             _ => return Err(NetavarkError::msg("Port forwarding tuples must contain only strings, encountered a non-string object")),
                         };
                         debug!("IP string from firewalld is {}", port_ip);
                         if port_ip != ipv4 && port_ip != ipv6 {
-                            port_forwarding_rules.append(port_tuple.try_clone().map_err(
-                                |e| NetavarkError::wrap("TODO: write a proper message", e.into()),
+                            port_forwarding_rules.append(wrap!(
+                                port_tuple.try_clone(),
+                                "Unable to clone port forwarding tuple "
                             )?)?;
                         }
                     }
```

Also note the use of the wrap!() macro which we have, that makes adding context to the error much easier.

Reply:
(It occurs to me that `wrap!` should be named `wrap_err!` or something)

Reply:
Maybe but renaming is out of scope for this PR